Std/Data/BitVec/Basic.lean

```diff
@@ -0,0 +1,185 @@
+  toString a := toString (repr a)
+
+/-- Theorem for normalizing the bit vector literal representation. -/
+@[simp] theorem ofNat_eq_ofNat : @OfNat.ofNat (BitVec n) i _ = BitVec.ofNat n i := rfl
+
+/-- Addition for bit vectors. -/
+protected def add (x y : BitVec n) : BitVec n :=
+  { val := x.val + y.val }
+/-- Subtraction for bit vectors. -/
+protected def sub (x y : BitVec n) : BitVec n :=
+  { val := x.val - y.val }
+/-- Multiplication for bit vectors. -/
+protected def mul (x y : BitVec n) : BitVec n :=
+  { val := x.val * y.val }
+/-- Modulo for bit vectors. -/
+protected def mod (x y : BitVec n) : BitVec n :=
+  { val := ⟨x.val % y.val, Nat.lt_of_le_of_lt (Nat.mod_le _ _) x.val.isLt⟩ }
+/-- Division for bit vectors. -/
+protected def div (x y : BitVec n) : BitVec n :=
+  { val := ⟨x.val / y.val, Nat.lt_of_le_of_lt (Nat.div_le_self _ _) x.val.isLt⟩ }
+/-- Less than for bit vectors. -/
+protected def lt (x y : BitVec n) : Bool :=
+  x.val < y.val
+/-- Less than or equal to for bit vectors. -/
+protected def le (x y : BitVec n) : Bool :=
+  x.val ≤ y.val
+
+instance : Add (BitVec n) := ⟨BitVec.add⟩
+instance : Sub (BitVec n) := ⟨BitVec.sub⟩
+instance : Mul (BitVec n) := ⟨BitVec.mul⟩
+instance : Mod (BitVec n) := ⟨BitVec.mod⟩
+instance : Div (BitVec n) := ⟨BitVec.div⟩
+instance : LT (BitVec n)  := ⟨fun x y => BitVec.lt x y⟩
+instance : LE (BitVec n)  := ⟨fun x y => BitVec.le x y⟩
+
+/-- Bitwise and for bit vectors. -/
+protected def and (x y : BitVec n) : BitVec n :=
+  { val := x.val &&& y.val }
+/-- Bitwise or for bit vectors. -/
+protected def or (x y : BitVec n) : BitVec n :=
+  { val := x.val ||| y.val }
+/-- Bitwise xor for bit vectors. -/
+protected def xor (x y : BitVec n) : BitVec n :=
+  { val := x.val ^^^ y.val }
+/-- Complement for bit vectors. -/
+protected def complement (x : BitVec n) : BitVec n :=
+  0 - (x + .ofNat n 1)
+/-- Shift left for bit vectors. -/
+protected def shiftLeft (a : BitVec n) (s : Nat) : BitVec n :=
+  .ofNat n (a.toNat <<< s)
+/-- Shift right for bit vectors. -/
+protected def shiftRight (a : BitVec n) (s : Nat) : BitVec n :=
+  .ofNat n (a.toNat >>> s)
+
+instance : Complement (BitVec w) := ⟨BitVec.complement⟩
+instance : AndOp (BitVec w) := ⟨BitVec.and⟩
+instance : OrOp (BitVec w) := ⟨BitVec.or⟩
+instance : Xor (BitVec w) := ⟨BitVec.xor⟩
+instance : HShiftLeft (BitVec w) Nat (BitVec w) := ⟨BitVec.shiftLeft⟩
+instance : HShiftRight (BitVec w) Nat (BitVec w) := ⟨BitVec.shiftRight⟩
+
+/-- Rotate left. -/
+def rotateLeft (x : BitVec w) (n : Nat) : BitVec w :=
+  x <<< n ||| x >>> (w - n)
+
+/-- Rotate right. -/
+def rotateRight (x : BitVec w) (n : Nat) : BitVec w :=
+  x >>> n ||| x <<< (w - n)
+
+/-- Append. -/
+def append (a : BitVec n) (b : BitVec m) : BitVec (n+m) :=
+  .ofNat (n+m) (a.toNat <<< m + b.toNat)
+
+instance : HAppend (BitVec w) (BitVec v) (BitVec (w+v)) := ⟨BitVec.append⟩
+
+/--
+Extraction of bits `i` down to `j` from a bit vector of size `n` to yield a
+new bitvector of size `i - j + 1`
```

Comment:
In mathlib4 we were leaning towards an API that satisfied
```suggestion
new bitvector of size `i - j`
```
As this saves some branching in `shrink`

Reply:
I imagine this does not align with SMTlib (and am aware it does not align with the convention in VHDL and Verilog), but I imagine it will be much more pleasant to prove things avout

Reply:
@eric-wieser Thanks for the feedback. Yes, the goal is to align with SMTlib which is a standard in software and hardware verification. SMT solvers are widely used in industry, and they all use the SMTlib standard.

Reply:
How about defining `extract` in terms of `extract'` using Eric's convention?

Reply:
`extract'` is a good idea. Please feel free to add a commit with its implementation.